drivers/SmartThings/jbl/src/jbl/device_manager.lua

```diff
@@ -0,0 +1,144 @@
+            audioTrackData.mediaSource = status.track.mediaSource.name
+        else
+            audioTrackData.mediaSource = ""
+        end
+
+        if is_new_audioTrackData(device, audioTrackData) then
+            log.info("device_manager.handle_status : update audioTrackData = " .. tostring(json.encode(audioTrackData)) .. ", dni = " .. tostring(device.device_network_id) .. tostring(json.encode(status)))
+            device:emit_event(capabilities.audioTrackData.audioTrackData(audioTrackData))
+            device:emit_event(capabilities.audioTrackData.totalTime(status.track.totalTime or 0))
+        end
+    end
+end
+
+
+function device_manager.update_status(driver, device)
+    local conn_info = device:get_field(fields.CONN_INFO)
+
+    if not conn_info then
+        log.warn(" device_manager.update_status : failed to find conn_info, dni = " .. tostring(device.device_network_id))
+        return
+    end
+
+    local response, err, status = conn_info:get_status()
+    if err or status ~= 200 then
+        log.error("device_manager.update_status : failed to get status, dni = " .. tostring(device.device_network_id) .. ", err = " .. tostring(err) .. ", status = " .. tostring(status))
+	if status == 404 then
+            log.info("device_manager.update_status : deleted, dni = " .. tostring(device.device_network_id))
+            device:offline()
+        end
+	return
+    end
+    device_manager.handle_status(driver, device, response)
+end
+
+local sse_event_handlers = {
+    ["message"] = device_manager.handle_status,
+}
+
+function device_manager.handle_sse_event(driver, device, event_type, data)
+    local status, device_json = pcall(json.decode, data)
+    if not status then
+        log.error(string.format("handle_sse_event : failed to decode data"))
+        return
+    end
+
+    local event_handler = sse_event_handlers[event_type]
+    if event_handler then
+        event_handler(driver, device, device_json)
+    else
+        log.error(string.format("handle_sse_event : unknown event type. dni = %s, event_type = '%s'", device.device_network_id, event_type))
+    end
+end
+
+function device_manager.refresh(driver, device)
+    device_manager.update_status(driver, device)
+end
+
+
+function device_manager.is_valid_connection(driver, device, conn_info)
+    if not conn_info then
+        log.error(" device_manager.is_valid_connection : failed to find conn_info, dni = " .. tostring(device.device_network_id))
+        return false
+    end
+    local _, err, status = conn_info:get_status()
+    if err or status ~= 200 then
+        log.error(" device_manager.is_valid_connection : failed to connect to device, dni = " .. tostring(device.device_network_id) .. ", err = " .. tostring(err) .. ", status = " .. tostring(status))
+        return false
+    end
+
+    return true
+end
+
+function device_manager.device_monitor(driver, device, device_info)
+    --TODO: add device monitoring logic (ip change, online/offline, etc ..)
```

Comment:
This seems like a pretty big todo, I just want to confirm it is still undone?

Reply:
Todo removed.